File: JavaAndScala/src/main/scala/org/lang/scala/conclusion/design/decorator/wrapper/SemicolonBufferedReader.scala
package org.lang.scala.conclusion.design.decorator.wrapper

import java.io.BufferedReader

/**
 * 	This is a stand-alone class to implement decorator pattern
 *  -- Decorator class inherits decorated class (i.e., BufferedReader) or interface
 *     -- Must declare an explicit constructor
 *        -- Need a "super" statement to make sure the compilation is not erroneous (the "super" statement is not actually working)
 * 	-- Maintain a reference of decorated class (i.e., BufferedReeader) inside decorator class (i.e., LineNumberDecorator)
 * 	   -- Use the explicit constructor to assign the decorated class instance to the reference: polymorphism
 * 	-- Override the method that needs to enhanced
 * 
 * 	@author VinceYuan
 */
class SemicolonBufferedReader(

    val bufferedReader: BufferedReader
    
) extends BufferedReader(bufferedReader) {
	
	/**
	 * 	Override the method that needs to be enhanced
	 */
	override def readLine() = {
	  
	  /*	Use decorated class to read a line	*/
	  var line = bufferedReader.readLine()
	  
	  /*	Match line with different cases	 */
	  line match {
	    /*	If line is null, then return null directly	*/
	    case null => null
	    /*	If line is not null, then add a semicolon at the end of a line and return	 */
	    case _ => {
	      line = line + ";"
				line
	    }
	  }
	}
}

File: JavaAndScala/src/main/java/com/java/se/conclusion/design/inheritance/wrapper/SemicolonBufferedReader.java
package com.java.se.conclusion.design.inheritance.wrapper;

import java.io.BufferedReader;
import java.io.Reader;

/**
 * 	This is a class to implement inheritance (to enhance a class)
 *  -- Typically this class is to enhance a method of a class
 * 	
 * 	To enhance a method of a class using inheritance: 
 *  -- Inherits the class & override the method that needs to be enhanced
 *  
 * @author VinceYuan
 *
 */
public class SemicolonBufferedReader extends BufferedReader {

	/*	Here inherits one of the parent constructors	*/
	public SemicolonBufferedReader(Reader in) {
		super(in);
	}
	
	/**
	 * 	Override the "readLine()" method for enhancement: 
	 * 	-- Aims to add the semicolon at the end of each line
	 */
	@Override
	public String readLine() {
		
		try {
			/*	If the line is null (i.e., read nothing from the file), then return null	*/
			String line = super.readLine();
			if (line == null) {
				return null;
				
			/*	Otherwise process what is read from reader	*/
			} else {
				line = line + ";";
				return line;	
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		
		/*	If there is any exception, return null	*/
		return null;
	}
}

